filebeat/input/file/state.go

```diff
@@ -56,83 +56,85 @@ func (s *State) IsEmpty() bool {
 
 // States handles list of FileState
 type States struct {
-	states []State
+	states map[string]State
 	sync.RWMutex
 }
 
 func NewStates() *States {
 	return &States{
-		states: []State{},
+		states: map[string]State{},
 	}
 }
 
 // Update updates a state. If previous state didn't exist, new one is created
 func (s *States) Update(newState State) {
+	s.UpdateWithTimestamp(newState, time.Now())
+}
+
+// UpdateWithTimestamp updates a state, using the passed timestamp. If the
+// previous state didn't exist, a new entry is created.
+func (s *States) UpdateWithTimestamp(newState State, ts time.Time) {
 	s.Lock()
 	defer s.Unlock()
 
-	index, _ := s.findPrevious(newState)
-	newState.Timestamp = time.Now()
+	// ensure ID is set, so it won't be generated in find and on insert
+	id := newState.ID()
```

Comment:
This reminds me of https://github.com/elastic/beats/issues/5547

Reply:
Somewhat related. The state is passed by value, while ID is tried to be generated lazily. Pass-by-value + lazy init does results in init multiple times...

Reply:
Do you think it's also something we should address in one of the other PR's?

Reply:
I'd address this in another PR. Question is: why is ID generated lazily, even though it's used right away for state lookup.

Reply:
I think there is not really a reason.